File: libreria_java/lib/src/main/java/com/example/DNIUtils.java
package com.example;

/**
 * Clase de utilidades para trabajar con DNI español.
 * Permite calcular la letra de un DNI y validar un DNI completo.
 */
public class DNIUtils {

    // Letras para el cálculo del DNI
    private static final char[] LETRAS = {
        'T', 'R', 'W', 'A', 'G', 'M', 'Y', 'F', 'P', 'D', 'X',
        'B', 'N', 'J', 'Z', 'S', 'Q', 'V', 'H', 'L', 'C', 'K', 'E'
    };

    /**
     * Calcula la letra correspondiente a un numero de DNI.
     *
     * @param numero Numero del DNI sin letra.
     * @return Letra del DNI correspondiente.
     * @throws IllegalArgumentException si el numero no esta entre 0 y 99_999_999
     */
    public static char calcularLetra(int numero) {
        if (numero < 0 || numero > 99999999) {
            throw new IllegalArgumentException("Numero de DNI fuera de rango");
        }
        return LETRAS[numero % 23];
    }

    /**
     * Valida un DNI completo (numero + letra).
     *
     * @param dni DNI completo (ejemplo: "12345678Z").
     * @return true si el DNI es valido; false en caso contrario.
     */
    public static boolean validarDni(String dni) {
        if (dni == null || dni.length() != 9) return false;

        String numeroStr = dni.substring(0, 8);
        char letra = Character.toUpperCase(dni.charAt(8));

        try {
            int numero = Integer.parseInt(numeroStr);
            return calcularLetra(numero) == letra;
        } catch (NumberFormatException e) {
            return false;
        }
    }
}

File: microservicio_python/app/dni_utils.py
LETRAS = "TRWAGMYFPDXBNJZSQVHLCKE"

def calcular_letra(numero: int) -> str:
    """Calcula la letra de un DNI a partir del número."""
    if numero <= 0 or numero > 99999999:
        raise ValueError("Número DNI fuera de rango")
    return LETRAS[numero % 23]

def validar_dni(dni: str) -> bool:
    """Valida un DNI completo (número + letra)."""
    if not dni or len(dni) != 9:
        return False

    numero_str = dni[:-1]
    letra = dni[-1].upper()

    if not numero_str.isdigit():
        return False

    numero = int(numero_str)
    letra_correcta = calcular_letra(numero)
    return letra == letra_correcta
